File: 15. Connect N ropes minimum cost.py
"""
Task: Connect N ropes with minimum cost
Time Complexity: O(n log n)
Space Complexity: O(n)
"""

import heapq

def min_cost(ropes):
    heapq.heapify(ropes)
    cost = 0
    while len(ropes) > 1:
        a = heapq.heappop(ropes)
        b = heapq.heappop(ropes)
        cost += a + b
        heapq.heappush(ropes, a+b)
    return cost

print('Min cost:', min_cost([4,3,2,6]))

File: 15. Connect N ropes minimum cost.cpp
/*
 Task: Connect N ropes with minimum cost
*/

#include <bits/stdc++.h>
using namespace std;

long long minCostConnectRopes(vector<int>& ropes)
{
    priority_queue<long long, vector<long long>, greater<long long>> pq;
    for (int r : ropes) pq.push(r);
    long long cost = 0;
    while (pq.size() > 1)
    {
        long long a = pq.top();
        pq.pop();
        long long b = pq.top();
        pq.pop();
        cost += a + b;
        pq.push(a + b);
    }
    return cost;
}

int main()
{
    vector<int> ropes = {4, 3, 2, 6};
    cout << "Min cost: " << minCostConnectRopes(ropes) << '\n';
    return 0;
}

// Time Complexity: O(n log n)
// Space Complexity: O(n)
